rust/arrow/src/array_equal.rs

```diff
@@ -0,0 +1,751 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::array::*;
+use crate::array_data::*;
+use crate::datatypes::*;
+use crate::util::bit_util;
+
+/// Trait for `Array` equality.
+pub trait ArrayEqual {
+    /// Returns true if this array is equal to the `other` array
+    fn equals(&self, other: &dyn Array) -> bool;
+
+    /// Returns true if the range [start_idx, end_idx) is equal to
+    /// [other_start_idx, other_start_idx + end_idx - start_idx) in the `other` array
+    fn range_equals(
+        &self,
+        other: &dyn Array,
+        start_idx: usize,
+        end_idx: usize,
+        other_start_idx: usize,
+    ) -> bool;
+}
+
+impl<T: ArrowPrimitiveType> ArrayEqual for PrimitiveArray<T> {
+    default fn equals(&self, other: &dyn Array) -> bool {
+        if !base_equal(&self.data(), &other.data()) {
+            return false;
+        }
+
+        let value_buf = self.data_ref().buffers()[0].clone();
+        let other_value_buf = other.data_ref().buffers()[0].clone();
+        let byte_width = T::get_bit_width() / 8;
+
+        if self.null_count() > 0 {
+            let values = value_buf.data();
+            let other_values = other_value_buf.data();
+
+            for i in 0..self.len() {
+                if self.is_valid(i) {
+                    let start = (i + self.offset()) * byte_width;
+                    let data = &values[start..(start + byte_width)];
+                    let other_start = (i + other.offset()) * byte_width;
+                    let other_data =
+                        &other_values[other_start..(other_start + byte_width)];
+                    if data != other_data {
+                        return false;
+                    }
+                }
+            }
+        } else {
+            let start = self.offset() * byte_width;
+            let other_start = other.offset() * byte_width;
+            let len = self.len() * byte_width;
+            let data = &value_buf.data()[start..(start + len)];
+            let other_data = &other_value_buf.data()[other_start..(other_start + len)];
+            if data != other_data {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    default fn range_equals(
+        &self,
+        other: &dyn Array,
+        start_idx: usize,
+        end_idx: usize,
+        other_start_idx: usize,
+    ) -> bool {
+        let other = other.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
```

Comment:
Should we check validity of index range at beginning of method so that we can make error message clearer?

Reply:
Good idea. I think we don't need to do much for `start_idx` and `end_idx`, but for `other_start_idx` we may want to check if it is out of range.